Periodically save state and store in indexedDB

kuychaco opened the pull request:
Fixes [atom/atom#4161](https://github.com/atom/atom/issues/4161)
- [x] separate serialization from package deactivation
- [x] extract serialization logic from unloadEditorWindow
- [x] fix package manager tests
- [x] save state asynchronously on keydown and mousedown, write tests
- [x] get rid of AtomEnvironment state instance variable
- [x] fix AtomEnvironment tests
- [x] profile serialization time for various environment states
- [x] store state in IndexedDB
- [x] add command-line flag to clear indexedDB object store

kuychaco committed:
Extract serialize functionality from ::deactivatePackage. Tests WIP

kuychaco committed:
Serialize in saveStateSync. Separate deserialization from deactivation.

kuychaco committed:
Fix tests for serialization

kuychaco committed:
Save state on mousedown or keypress events (debounce 1s). Tests WIP

kuychaco commented on spec/atom-environment-spec.coffee:
```diff
@@ -174,6 +174,19 @@ describe "AtomEnvironment", ->
+      atom.document.dispatchEvent(mousedown)
+      advanceClock 1100
+      expect(atom.saveStateSync).toHaveBeenCalled()
+
```
Spy not working as expected and causing test to fail:

<img width="794" alt="screen shot 2016-01-27 at 11 18 09 am" src="https://cloud.githubusercontent.com/assets/7910250/12623578/0f524114-c4e8-11e5-8ffe-a0d7a774c6cc.png">

kuychaco replied:
Figured out this was an issue with binding `saveStateSync` before spying on it

nathansobo commented on spec/package-manager-spec.coffee:
```diff
@@ -821,6 +819,35 @@ describe "PackageManager", ->
           expect(addErrorHandler.callCount).toBe 0
 
+  describe "serialization", ->
+
```
Super minor nitpick drive-by comment: :fire: this empty line. :smile:

jeancroy commented:
Is this saved to .atom/storage ?
On the forum, when atom crash on startup, deleting the storage folder is a standard debug step.
Maybe there's an interest in discarding most of the state but restoring unsaved file content.

kuychaco committed:
Avoid binding method before it’s spied upon when debouncing

kuychaco committed:
:art: clean up test

kuychaco committed:
Assign debounce interval to property

kuychaco commented:
@jeancroy thanks for your comment. To answer your question - yes, the atom environment state is saved to .atom/storage. That said, I'm going to create an unsaved text buffers store that will be located elsewhere (possibly using IndexedDB). So in the event of a crash and deletion of the storage folder, your unsaved buffers will be preserved. 

The unsaved buffer store will be a part of a separate pull request. I've updated the title of this one to be more specific as to what aspect of crash recovery will be addressed here.

maxbrunsfeld commented:
This is looking good! How long does `saveStateSync` on average? Does it vary much depending on how many open buffers you have?

kuychaco committed:
Rename StorageFolder::store to ::storeSync

kuychaco committed:
Save state on keydown rather than on keypress

kuychaco committed:
Save state asynchronously on mousedown and keydown events. Tests WIP

kuychaco committed:
Remove AtomEnvironment state instance variable. Tests WIP

YuriSolovyov commented:
I wonder if we can use [requestIdleCallback](https://developers.google.com/web/updates/2015/08/using-requestidlecallback) here once we have chrome 47+

maxbrunsfeld commented:
> I wonder if we can use requestIdleCallback here once we have chrome 47+

Yeah, we should start using that, as soon as https://github.com/atom/atom/pull/9627 lands

nathansobo commented on src/storage-folder.coffee:
```diff
@@ -6,11 +6,16 @@ class StorageFolder
 
     fs.writeFileSync(@pathForKey(name), JSON.stringify(object), 'utf8')
 
+  storeAsync: (name, object) ->
```
I think this should be called `store`... the convention in Node is to have async be implied and only mention sync vs async on a sync alternative.

kuychaco committed:
Use  method name `store` rather than `storeAsync`

kuychaco committed:
Fix tests

kuychaco committed:
Add env version and window dimensions to serialized state

kuychaco committed:
Call AtomEnvironment::deserialize in AtomEnvironment::loadStateSync

kuychaco committed:
Remove atom environment state instance variable

kuychaco committed:
Add disposable for removing debouncedSaveState listener

kuychaco committed:
Serialize package upon deactivation

kuychaco committed:
Only call package serialize methods once on quit

kuychaco commented:
Did some profiling. Here are durations for saveState (synchronous serialization, and async writing to disk):

| State | Time |
| ------- | ------- |
| 20 tabs opened | 2 ms |
| large unsaved buffer (jquery, 10K lines)* | 11 ms |

*Marker Layer serialization took 8.3 ms

20 tabs open:
![screen shot 2016-02-01 at 5 58 09 pm](https://cloud.githubusercontent.com/assets/7910250/12737447/6c3014c8-c90d-11e5-8863-23956a9aeb6a.png)

large unsaved buffer (jquery, 10K lines)
![screen shot 2016-02-01 at 5 16 20 pm](https://cloud.githubusercontent.com/assets/7910250/12737350/6c9c0ea4-c90c-11e5-9b1d-4f9b15927823.png)

as-cii commented:
Great job, @kuychaco. I think those are really impressive results: the ideal idle time is `50ms`, which, if I am interpreting the results correctly, means we have room for serializing ~ 5 large files or ~ 400 open tabs.

Also, I find very interesting that most of the time is spent in dumping markers: I guess it's because we have to perform a full traversal of the tree, which can be time consuming for large files. I still wonder, though, if we can dissect that `MarkerIndex` call and maybe optimize based on that.

Also, @kuychaco, what's the time spent in `JSON.stringify`?

I am really excited about this! :sparkles:

kuychaco committed:
Create StateStore class and specs for storing state in IndexedDB

kuychaco committed:
Refactor to use StateStore instead of StorageFolder

kuychaco committed:
Add storedAt date for serialized environment state

nathansobo commented:
> Also, I find very interesting that most of the time is spent in dumping markers

I'd like to implement serialization of the `MarkerIndex` directly for this case. Then we don't have to traverse the tree and build an alternate structure. Instead we can just convert the tree to JSON directly.

kuychaco committed:
Add ‘indexedDB’ to list of known global variables

kuychaco committed:
Use prerelease version of text-buffer

kuychaco committed:
Remove empty parameter list in order to pass linter

kuychaco committed:
Merge branch 'master' into ku-crash-recovery

maxbrunsfeld committed:
Don't emit window:loaded event until async window initialization completes

Signed-off-by: Katrina Uychaco <kuychaco@github.com>

maxbrunsfeld committed:
Handle database connection errors gracefully

When opening a second Atom instance (e.g. when running the integration specs)
indexedDB connections will fail. In this case, StateStore.prototype.save
and StateStore.prototype.load will become noops.

Signed-off-by: Katrina Uychaco <kuychaco@github.com>

maxbrunsfeld committed:
Add --user-data-dir flag, to control indexedDB directory

This way, we can still use indexedDB in the integration tests

Signed-off-by: Katrina Uychaco <kuychaco@github.com>

maxbrunsfeld committed:
:shirt: Use single quotes in JS string

Signed-off-by: Katrina Uychaco <kuychaco@github.com>

kuychaco committed:
Add test to check for indexedDB connection

kuychaco committed:
Merge branch 'master' into ku-crash-recovery

kuychaco committed:
Use text-buffer 8.2.2-2

kuychaco committed:
Fix bug from merge with PR#10743

kuychaco committed:
Wait for tab to open in webdriverio `waitForPaneItemCount` command

kuychaco committed:
Use `user-data-dir` with temp directory when running core tests

This is to ensure successful database connection. Since core specs and 
package specs are run at the same time and both open an indexedDB 
connection, there were occasional core spec failures due to failed
database connection.

kuychaco committed:
In startup-spec check pane item count after editor has been loaded

kuychaco committed:
Revert "Wait for tab to open in webdriverio `waitForPaneItemCount` command"

This reverts commit 8e33f8bf5b974a690cf291dcfad5a38b5cc12a57.

kuychaco committed:
Revert test since `startEditorWindow` no longer returns a promise

kuychaco commented:
/cc @atom/core feedback welcome. Note that the travis failure can be disregarded ([flakey spec](https://github.com/atom/atom/issues/10435))

This PR is intended to improve crash recovery by serializing and saving the AtomEnvironment state regularly to minimize data loss.

**Serializing**
Serializing will occur upon debounced mousedown and keydown events (debouce interval is set to 1 second). As part of this PR, I've cleaned up the `AtomEnvironment` code and done away with the `state` instance variable. Assuming that the serialized state is successfully saved, that means that worst-case-scenario a lovely Atom user will lose only 1 second worth of work upon crash. And speaking of successful saving...

**Saving**
With this PR, the state is now stored in IndexedDB rather than the ~/.atom/storage folder. Here are some of the benefits of using IndexedDB:
- transactional nature means:
  - when the disk is full we won't lose data as we would in the case of truncated files on disk
  - when two Atom windows have the same project folders open we don't have to worry about race conditions
- IndexedDB can store JS objects so there is no need to JSON.stringify/parse the state (possible performance improvements for startup time if deserializing large state object)
- state can be easily accessed in chrome dev tools, which is pretty cool I think
 ![screen shot 2016-02-11 at 5 36 13 pm](https://cloud.githubusercontent.com/assets/7910250/12996586/0620e816-d0e6-11e5-8e95-36900898a036.png)

TODO: come up with a user-friendly way to clear the state from IndexedDB for [debugging purposes](https://github.com/atom/atom/issues/2957)

kuychaco committed:
Add ability to clear IndexedDB state object store

kuychaco committed:
Add command-line argument to clear state in IndexedDB

nathansobo commented on src/atom-environment.coffee:
```diff
@@ -640,7 +648,7 @@ class AtomEnvironment extends Model
 
     @packages.loadPackages()
-
+    
```
:fire: this trailing whitespace

nathansobo commented on src/storage-folder.coffee:
```diff
@@ -6,11 +6,16 @@ class StorageFolder
 
     fs.writeFileSync(@pathForKey(name), JSON.stringify(object), 'utf8')
 
+  store: (name, object) ->
```
Is this async version used now that you're using IndexedDB?

nathansobo commented:
Very solid work @kuychaco. I did a pretty quick scan and things look good to me. Are you thinking of adding the `--clear-workspace-state` flag before merging?

kuychaco committed:
:fire: trailing white space

kuychaco committed:
Remove unused StorageFolder::store method (moved state to IndexedDB)

kuychaco commented:
@nathansobo thanks for the review! I think I will add the command-line flag before merging. Hope to get that done tonight.

kuychaco committed:
Clear state in indexedDB if ‘clear-state’ command-line flag is passed

kuychaco committed:
Merge branch 'ku-clear-indexedDB' into ku-crash-recovery

kuychaco committed:
Merge branch 'master' into ku-crash-recovery

kuychaco committed:
Use pre-released text-buffer version 8.3.1-0

kuychaco committed:
Use more descriptive flag for clearing state - ‘clear-window-state’

kuychaco committed:
:arrow_up: text-buffer

benogle commented:
:boom:

jeancroy commented:
Hi @kuychaco 

Is there mechanism to migrate state data from previous format to the new IndexedDB format ?
Here's one thread where a beta user think he lost all his state data on upgrade, but it was there, just ignored/not loaded.

https://discuss.atom.io/t/upgrading-to-1-7-0-beta0-blew-away-my-session-any-way-to-restore/26918
It could be interesting to avoid a widespread fear of data loss.

iolsen commented:
cc @as-cii ☝️

nathansobo commented:
Talked with @as-cii and he's planning on falling back to reading the state out of `~/.atom` if it is present and there's nothing in the database.

mehcode commented:
Great to hear there is a fallback in the works. :+1: 

---

> [...] People care more about their session state than I realized.

People (me) have even written a [package](https://github.com/mehcode/atom-project-plus) to list / inspect / make use of session state. It really is a nice feature of Atom.

miguelbarrio reviewed (commented):
8d55bbcdeadc41b07a1e5e710e45b4a939c399eb
